pkg/registry/core/service/ipallocator/ipallocator.go

```diff
@@ -0,0 +1,480 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipallocator
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/netip"
+	"time"
+
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	networkingv1alpha1informers "k8s.io/client-go/informers/networking/v1alpha1"
+	networkingv1alpha1client "k8s.io/client-go/kubernetes/typed/networking/v1alpha1"
+	networkingv1alpha1listers "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	netutils "k8s.io/utils/net"
+)
+
+const ControllerName = "ipallocator.k8s.io"
+
+// Allocator implements current ipallocator interface using IPAddress API object
+// and an informer as backend.
+type Allocator struct {
+	net          *net.IPNet
+	firstAddress netip.Addr   // first IP address within the range
+	lastAddress  netip.Addr   // last IP address within the range
+	family       api.IPFamily // family is the IP family of this range
+	max          int64        // cap the total number of IPs available to maxInt64
+	offset       int          // subdivides the assigned IP range to prefer dynamic allocation from the upper range
+	rand         *rand.Rand
+
+	client          networkingv1alpha1client.NetworkingV1alpha1Interface
+	ipAddressLister networkingv1alpha1listers.IPAddressLister
+	ipAddressSynced cache.InformerSynced
+
+	// metrics is a metrics recorder that can be disabled
+	metrics metricsRecorderInterface
+}
+
+var _ Interface = &Allocator{}
+
+// NewIPAllocator returns an IP allocator that use the IPAddress object
+// to track the assigned IP addresses, using an informer cache as storage.
+func NewIPAllocator(
+	cidr *net.IPNet,
+	client networkingv1alpha1client.NetworkingV1alpha1Interface,
+	ipAddressInformer networkingv1alpha1informers.IPAddressInformer,
+) (*Allocator, error) {
+	prefix, err := netip.ParsePrefix(cidr.String())
+	if err != nil {
+		return nil, err
+	}
+
+	if prefix.Addr().Is6() && prefix.Bits() < 64 {
+		return nil, fmt.Errorf("maximum service CIDR is /64 , got /%d", prefix.Bits())
+	}
+
+	max := netutils.RangeSize(cidr)
+	// Don't use the network's ".0" address, but don't just Allocate() it - we
+	// don't ever want to be able to release it.
+	ipFirst := prefix.Masked().Addr().Next()
+	max--
+
+	var family api.IPFamily
+	ipLast := broadcastAddress(prefix)
+	if netutils.IsIPv6CIDR(cidr) {
+		family = api.IPv6Protocol
+	} else {
+		family = api.IPv4Protocol
+		// Don't use the IPv4 network's broadcast address, but don't just
+		// Allocate() it - we don't ever want to be able to release it.
+		ipLast = ipLast.Prev()
```

Comment:
wait for @thockin , he is challenging we should use the zero and broadcast address too ... I had to look it and he is right XD 

```
RFC 1878

*Subnet all zeroes and all ones excluded. (Obsolete)
*Host all zeroes and all ones excluded.   (Obsolete)
```

Reply:
please don't - while in theory it will work (since the address is rewritten in proxy before the packet ever leave the host stack) it will confuse anybody coming at it from networking side

Reply:
😄